README.md
# Least Recently Used (LRU) cache algorithm

A finite key-value cache using the [Least Recently Used (LRU)](http://en.wikipedia.org/wiki/Cache_algorithms#Least_Recently_Used) cache algorithm where the most recently used objects are keept in cache while less recently used items are purged.

This implementation is compatible with most JavaScript environments (including ye olde browser) and is very efficient.

## Terminology & design

- Based on a doubly-linked list for low complexity random shuffling of entries.

- The cache object iself has a "head" (least recently used entry) and a
  "tail" (most recently used entry).

- The "head" and "tail" are "entries" -- an entry might have a "newer" and
  an "older" entry (doubly-linked, "older" being close to "head" and "newer"
  being closer to "tail").

- Key lookup is done through a key-entry mapping native object, which on most
  platforms mean `O(1)` complexity. This comes at a very low memory cost  (for
  storing two extra pointers for each entry).

Fancy ASCII art illustration of the general design:

        entry             entry             entry             entry        
        ______            ______            ______            ______       
       | head |.newer => |      |.newer => |      |.newer => | tail |      
       |  A   |          |  B   |          |  C   |          |  D   |      
       |______| <= older.|______| <= older.|______| <= older.|______|      

    removed  <--  <--  <--  <--  <--  <--  <--  <--  <--  <--  <--  added

## Example

    var c = new LRUCache(3);
    c.put('adam', 29);
    c.put('john', 26);
    c.put('angela', 24);
    c.toString();        // -> "adam:29 < john:26 < angela:24"
    c.get('john');       // -> 26
    // Now 'john' is the most recently used entry, since we just requested it
    c.toString();        // -> "adam:29 < angela:24 < john:26"
    c.put('zorro', 141); // -> {key:adam, value:29}
    // Because we only have room for 3 entries, put-ing 'zorro' purged 'adam'
    c.toString();        // -> "angela:24 < john:26 < zorro:141"

# API

An entry is a simple `Object` with at least two members: `{key:Object, value:Object}`. An entry might also have a `newer` member which points to a newer entry, and/or a `older` member pointing to an older entry.

## Included in lru_cache and lru_cache/core:

### new LRUCache(Number limit) -> LRUCache instance

Creates a new cache object which will hold up to `limit` entries.

### *LRUCache.prototype*.size -> Number

Current number of entries. Read-only.

### *LRUCache.prototype*.limit <-> Number

Maximum number of items this cache will keep.

### *LRUCache.prototype*.put (Object key, Object value) -> Object entry

Put `value` into the cache associated with `key`.

**Returns an entry which was removed** (to make room for the new entry) or  `undefined` if there was enough space for the new entry.

> **Note:** The returned entry does **not** include any (strong) references to other entries (i.e. there is no `older` or `newer` members). This design makes garbage collection predictable.

### *LRUCache.prototype*.get (Object key) -> Object value

Retrieve value for, and register recent use of, `key`. Returns the value associated with `key` or `undefined` if not in the cache.

### *LRUCache.prototype*.shift () -> Object entry

Remove the least recently used (oldest) entry. Returns the removed entry, or `undefined` if the cache was empty.

If you need to perform any form of finalization of purged items, this is a good place to do it. Simply override/replace this function:

    var c = new LRUCache(123);
    c.shift = function() {
      var entry = LRUCache.prototype.shift.call(this);
      doSomethingWith(entry);
      return entry;
    }

The returned entry must not include any strong references to other entries. See note in the documentation of `LRUCache.prototype.put (Object key, Object value) -> Object entry`.

## Included in lru_cache only

### *LRUCache.prototype*.find (Object key) -> Object entry

Check if `key` is in the cache *without registering recent use*. Feasible if
you do not want to chage the state of the cache, but only "peek" at it.
Returns the entry associated with `key` if found, otherwise `undefined` is
returned.

> **Note:** The entry returned is *managed by the cache* (until purged) and thus contains members with strong references which might be altered at any time by the cache object. You should look at the returned entry as being immutable.

### *LRUCache.prototype*.set (key, value) -> Object oldValue

Update the value of entry with `key` or `put` a new entry. Returns the old value, or undefined if the cache was empty.

### *LRUCache.prototype*.remove (key) -> Object value

Remove entry `key` from cache and return its value. Returns `undefined` if `key` is not found.

### *LRUCache.prototype*.removeAll () -> LRUCache instance

Removes all entries and return itself.

### *LRUCache.prototype*.keys () -> Array keys

Return an array containing all keys of entries in arbitrary order.

### *LRUCache.prototype*.forEach (fun, [Object context, Boolean desc | true])

Call `fun` for each entry. Starting with the newest entry if `desc` is a true
value, otherwise starts with the oldest (head) enrty and moves towards the
tail.

Returns nothing (`undefined`).

`fun` is called with 3 arguments in the context `context`:

    fun.call(context, Object key, Object value, LRUCache self)

Example which prints "key: value" starting with the most recent entry:

    cache.forEach(function(key, value) {
      puts(key+': '+value);
    }, true);


### *LRUCache.prototype*.toJSON () -> Array representation

Returns an array of object (for use by `JSON.stringify`) of the form:

    [
      {key:"key1", value:"value1"},
      {key:"key2", value:"value2"},
      {key:"key3", value:"value3"}
    ]

### *LRUCache.prototype*.toString () -> String representation

Returns a string representation in the format:

    key1:value1 < key2:value2 < key3:value3

Oldest (head) on the left hand side and newer entries to the right hand side.

## Factorising a minimal implementation

As this code is most suitable for embedding, here is a shortlist of the essential parts (prototype functions) needed for a minimal implementation. All other functions, not mentioned here, are simply ancillary.

- **LRUCache** -- the constructor is naturally a good thing to keep ;)
- *LRUCache.prototype*.**put** -- handles appending and chaining.
- *LRUCache.prototype*.**shift** -- used by **put** to "purge" an old entry.
- *LRUCache.prototype*.**get** -- fetches a cached entry and registers that entry as being recently used.

To include only the minimal code above, require lru_cache/core instead of lru_cache

# MIT license

Copyright (c) 2016 Rasmus Andersson <http://hunch.se/>, Ben Woosley

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG
Version 1.0.2 (Released 2016-03-29)
===================================
* Update copyright to 2016
* Update some docs references to js-lru and lru-core.js

Version 1.0.1 (Released 2016-03-29)
===================================
* Add links to Changelog
* Rename lru-core to core.js, for inclusion via require('lru_cache/core')
* Rename npm package and github to lru_cache, to match the main object
https://github.com/Empact/lru_cache
http://npmjs.com/package/lru_cache

Version 1.0.0 (Released 2016-03-29)
===================================
Bugs
* Fix that shift didn't update size
* Fix that shift didn't properly handle the last element
https://github.com/rsms/js-lru/issues/6

Features
* Enable the user to require lru-core, which is a minimal set of functionality
* Publish to npm under lru-js
https://github.com/rsms/js-lru/issues/7


test.js
// Test which will run in nodejs
// $ node test.js
// (Might work with other CommonJS-compatible environments)
'use strict'

var assert = require('assert')
var LRUCache = require('./index').LRUCache
var c = new LRUCache(4)

c.put('adam', 29)
c.put('john', 26)
c.put('angela', 24)
c.put('bob', 48)
assert.equal(c.toString(), 'adam:29 < john:26 < angela:24 < bob:48')

assert.equal(c.get('adam'), 29)
assert.equal(c.get('john'), 26)
assert.equal(c.get('angela'), 24)
assert.equal(c.get('bob'), 48)
assert.equal(c.toString(), 'adam:29 < john:26 < angela:24 < bob:48')

assert.equal(c.get('angela'), 24)
assert.equal(c.toString(), 'adam:29 < john:26 < bob:48 < angela:24')

c.put('ygwie', 81)
assert.equal(c.toString(), 'john:26 < bob:48 < angela:24 < ygwie:81')
assert.equal(c.get('adam'), undefined)

c.set('john', 11)
assert.equal(c.toString(), 'bob:48 < angela:24 < ygwie:81 < john:11')
assert.equal(c.get('john'), 11)

var expectedKeys = ['bob', 'angela', 'ygwie', 'john']
c.forEach(function (k, v) {
  // sys.puts(k+': '+v)
  assert.equal(k, expectedKeys.shift())
})

var current_size = c.size
c.remove('john')
assert.equal(current_size - 1, c.size)

// test remove
var to_remove = new LRUCache(4)
to_remove.put('adam', 29)
to_remove.put('john', 26)
to_remove.remove('adam')
to_remove.remove('john')
assert.equal(to_remove.size, 0)
assert.equal(to_remove.head, undefined)
assert.equal(to_remove.tail, undefined)

// test shift
c = new LRUCache(4)
assert.equal(c.size, 0)
c.put('a', 1)
c.put('b', 2)
c.put('c', 3)
assert.equal(c.size, 3)
var e = c.shift()
assert.equal(e.key, 'a')
assert.equal(e.value, 1)
e = c.shift()
assert.equal(e.key, 'b')
assert.equal(e.value, 2)
e = c.shift()
assert.equal(e.key, 'c')
assert.equal(e.value, 3)
// test that forEach does not execute against any entries
c.forEach(function () { assert(false) }, true)
assert.equal(c.size, 0)

c = new LRUCache(4)

c.put(0, 1)
c.put(0, 2)
c.put(0, 3)
c.put(0, 4)

assert.equal(c.size, 4)
assert.deepEqual(c.shift(), { key: 0, value: 1, newer: undefined, older: undefined })
assert.deepEqual(c.shift(), { key: 0, value: 2, newer: undefined, older: undefined })
assert.deepEqual(c.shift(), { key: 0, value: 3, newer: undefined, older: undefined })
assert.deepEqual(c.shift(), { key: 0, value: 4, newer: undefined, older: undefined })
assert.equal(c.size, 0) // check .size correct
c.forEach(function () { assert(false) }, undefined, true)  // check .tail correct
// If we made it down here, all tests passed. Neat.
console.log('All tests passed!')


lru.js
/**
 * A doubly linked list-based Least Recently Used (LRU) cache. Will keep most
 * recently used items while discarding least recently used items when its limit
 * is reached.
 *
 * Licensed under MIT. Copyright (c) 2010 Rasmus Andersson <http://hunch.se/>
 * See README.md for details.
 *
 * Illustration of the design:
 *
 *       entry             entry             entry             entry
 *       ______            ______            ______            ______
 *      | head |.newer => |      |.newer => |      |.newer => | tail |
 *      |  A   |          |  B   |          |  C   |          |  D   |
 *      |______| <= older.|______| <= older.|______| <= older.|______|
 *
 *  removed  <--  <--  <--  <--  <--  <--  <--  <--  <--  <--  <--  added
 */
function LRUCache (limit) {
  // Current size of the cache. (Read-only).
  this.size = 0
  // Maximum number of items this cache can hold.
  this.limit = limit
  this._keymap = {}
}

/**
 * Put <value> into the cache associated with <key>. Returns the entry which was
 * removed to make room for the new entry. Otherwise undefined is returned
 * (i.e. if there was enough room already).
 */
LRUCache.prototype.put = function (key, value) {
  var entry = {key: key, value: value}
  // Note: No protection agains replacing, and thus orphan entries. By design.
  this._keymap[key] = entry
  if (this.tail) {
    // link previous tail to the new tail (entry)
    this.tail.newer = entry
    entry.older = this.tail
  } else {
    // we're first in -- yay
    this.head = entry
  }
  // add new entry to the end of the linked list -- it's now the freshest entry.
  this.tail = entry
  if (this.size === this.limit) {
    // we hit the limit -- remove the head
    return this.shift()
  } else {
    // increase the size counter
    this.size++
  }
}

/**
 * Purge the least recently used (oldest) entry from the cache. Returns the
 * removed entry or undefined if the cache was empty.
 *
 * If you need to perform any form of finalization of purged items, this is a
 * good place to do it. Simply override/replace this function:
 *
 *   var c = new LRUCache(123);
 *   c.shift = function() {
 *     var entry = LRUCache.prototype.shift.call(this);
 *     doSomethingWith(entry);
 *     return entry;
 *   }
 */
LRUCache.prototype.shift = function () {
  // todo: handle special case when limit == 1
  var entry = this.head
  if (entry) {
    if (this.head.newer) {
      // advance the list
      this.head = this.head.newer
      this.head.older = undefined
    } else {
      // the cache is exhausted
      this.head = undefined
      this.tail = undefined
    }
    // Remove last strong reference to <entry> and remove links from the purged
    // entry being returned:
    entry.newer = entry.older = undefined
    // delete is slow, but we need to do this to avoid uncontrollable growth:
    delete this._keymap[entry.key]
    this.size--
  }
  return entry
}

/**
 * Get and register recent use of <key>. Returns the value associated with <key>
 * or undefined if not in cache.
 */
LRUCache.prototype.get = function (key, returnEntry) {
  // First, find our cache entry
  var entry = this._keymap[key]
  if (entry === undefined) {
    return // Not cached. Sorry.
  }
  // As <key> was found in the cache, register it as being requested recently
  if (entry === this.tail) {
    // Already the most recenlty used entry, so no need to update the list
    return returnEntry ? entry : entry.value
  }
  // HEAD--------------TAIL
  //   <.older   .newer>
  //  <--- add direction --
  //   A  B  C  <D>  E
  if (entry.newer) {
    if (entry === this.head) {
      this.head = entry.newer
    }
    entry.newer.older = entry.older // C <-- E.
  }
  if (entry.older) {
    entry.older.newer = entry.newer // C. --> E
  }
  entry.newer = undefined // D --x
  entry.older = this.tail // D. --> E
  if (this.tail) {
    this.tail.newer = entry // E. <-- D
  }
  this.tail = entry
  return returnEntry ? entry : entry.value
}

// ----------------------------------------------------------------------------
// Following code is optional and can be removed without breaking the core
// functionality.

/**
 * Check if <key> is in the cache without registering recent use. Feasible if
 * you do not want to chage the state of the cache, but only "peek" at it.
 * Returns the entry associated with <key> if found, or undefined if not found.
 */
LRUCache.prototype.find = function (key) {
  return this._keymap[key]
}

/**
 * Update the value of entry with <key>. Returns the old value, or undefined if
 * entry was not in the cache.
 */
LRUCache.prototype.set = function (key, value) {
  var oldvalue
  var entry = this.get(key, true)
  if (entry) {
    oldvalue = entry.value
    entry.value = value
  } else {
    oldvalue = this.put(key, value)
    if (oldvalue) oldvalue = oldvalue.value
  }
  return oldvalue
}

/**
 * Remove entry <key> from cache and return its value. Returns undefined if not
 * found.
 */
LRUCache.prototype.remove = function (key) {
  var entry = this._keymap[key]
  if (!entry) {
    return
  }
  delete this._keymap[entry.key] // need to do delete unfortunately
  if (entry.newer && entry.older) {
    // relink the older entry with the newer entry
    entry.older.newer = entry.newer
    entry.newer.older = entry.older
  } else if (entry.newer) {
    // remove the link to us
    entry.newer.older = undefined
    // link the newer entry to head
    this.head = entry.newer
  } else if (entry.older) {
    // remove the link to us
    entry.older.newer = undefined
    // link the newer entry to head
    this.tail = entry.older
  } else {// if(entry.older === undefined && entry.newer === undefined) {
    this.head = this.tail = undefined
  }

  this.size--
  return entry.value
}

/** Removes all entries */
LRUCache.prototype.removeAll = function () {
  // This should be safe, as we never expose strong refrences to the outside
  this.head = this.tail = undefined
  this.size = 0
  this._keymap = {}
}

/**
 * Return an array containing all keys of entries stored in the cache object, in
 * arbitrary order.
 */
if (typeof Object.keys === 'function') {
  LRUCache.prototype.keys = function () {
    return Object.keys(this._keymap)
  }
} else {
  LRUCache.prototype.keys = function () {
    var keys = []
    for (var k in this._keymap) keys.push(k)
    return keys
  }
}

/**
 * Call `fun` for each entry. Starting with the newest entry if `desc` is a true
 * value, otherwise starts with the oldest (head) enrty and moves towards the
 * tail.
 *
 * `fun` is called with 3 arguments in the context `context`:
 *   `fun.call(context, Object key, Object value, LRUCache self)`
 */
LRUCache.prototype.forEach = function (fun, context, desc) {
  var entry
  if (context === true) {
    desc = true
    context = undefined
  } else if (typeof context !== 'object') {
    context = this
  }
  if (desc) {
    entry = this.tail
    while (entry) {
      fun.call(context, entry.key, entry.value, this)
      entry = entry.older
    }
  } else {
    entry = this.head
    while (entry) {
      fun.call(context, entry.key, entry.value, this)
      entry = entry.newer
    }
  }
}

/** Returns a JSON (array) representation */
LRUCache.prototype.toJSON = function () {
  var s = []
  var entry = this.head
  while (entry) {
    s.push({key: entry.key.toJSON(), value: entry.value.toJSON()})
    entry = entry.newer
  }
  return s
}

/** Returns a String representation */
LRUCache.prototype.toString = function () {
  var s = ''
  var entry = this.head
  while (entry) {
    s += String(entry.key) + ':' + entry.value
    entry = entry.newer
    if (entry) {
      s += ' < '
    }
  }
  return s
}

// Export ourselves
if (typeof this === 'object') {
  this.LRUCache = LRUCache
}


core.js
/**
 * A doubly linked list-based Least Recently Used (LRU) cache. Will keep most
 * recently used items while discarding least recently used items when its limit
 * is reached.
 *
 * Licensed under MIT. Copyright (c) 2016 Rasmus Andersson <http://hunch.se/>,
 * Ben Woosley. See README.md for details.
 *
 * Illustration of the design:
 *
 *       entry             entry             entry             entry
 *       ______            ______            ______            ______
 *      | head |.newer => |      |.newer => |      |.newer => | tail |
 *      |  A   |          |  B   |          |  C   |          |  D   |
 *      |______| <= older.|______| <= older.|______| <= older.|______|
 *
 *  removed  <--  <--  <--  <--  <--  <--  <--  <--  <--  <--  <--  added
 */
'use strict'

function LRUCache (limit) {
  // Current size of the cache. (Read-only).
  this.size = 0
  // Maximum number of items this cache can hold.
  this.limit = limit
  this.head = this.tail = undefined
  this._keymap = {}
}

/**
 * Put <value> into the cache associated with <key>. Returns the entry which was
 * removed to make room for the new entry. Otherwise undefined is returned
 * (i.e. if there was enough room already).
 */
LRUCache.prototype.put = function (key, value) {
  var entry = {key: key, value: value}
  // Note: No protection agains replacing, and thus orphan entries. By design.
  this._keymap[key] = entry
  if (this.tail) {
    // link previous tail to the new tail (entry)
    this.tail.newer = entry
    entry.older = this.tail
  } else {
    // we're first in -- yay
    this.head = entry
  }
  // add new entry to the end of the linked list -- it's now the freshest entry.
  this.tail = entry
  if (this.size === this.limit) {
    // we hit the limit -- remove the head
    return this.shift()
  } else {
    // increase the size counter
    this.size++
  }
}

/**
 * Purge the least recently used (oldest) entry from the cache. Returns the
 * removed entry or undefined if the cache was empty.
 *
 * If you need to perform any form of finalization of purged items, this is a
 * good place to do it. Simply override/replace this function:
 *
 *   var c = new LRUCache(123);
 *   c.shift = function() {
 *     var entry = LRUCache.prototype.shift.call(this);
 *     doSomethingWith(entry);
 *     return entry;
 *   }
 */
LRUCache.prototype.shift = function () {
  // todo: handle special case when limit == 1
  var entry = this.head
  if (entry) {
    if (this.head.newer) {
      // advance the list
      this.head = this.head.newer
      this.head.older = undefined
    } else {
      // the cache is exhausted
      this.head = undefined
      this.tail = undefined
    }
    // Remove last strong reference to <entry> and remove links from the purged
    // entry being returned:
    entry.newer = entry.older = undefined
    // delete is slow, but we need to do this to avoid uncontrollable growth:
    delete this._keymap[entry.key]
    this.size--
  }
  return entry
}

/**
 * Get and register recent use of <key>. Returns the value associated with <key>
 * or undefined if not in cache.
 */
LRUCache.prototype.get = function (key, returnEntry) {
  // First, find our cache entry
  var entry = this._keymap[key]
  if (entry === undefined) {
    return // Not cached. Sorry.
  }
  // As <key> was found in the cache, register it as being requested recently
  if (entry === this.tail) {
    // Already the most recenlty used entry, so no need to update the list
    return returnEntry ? entry : entry.value
  }
  // HEAD--------------TAIL
  //   <.older   .newer>
  //  <--- add direction --
  //   A  B  C  <D>  E
  if (entry.newer) {
    if (entry === this.head) {
      this.head = entry.newer
    }
    entry.newer.older = entry.older // C <-- E.
  }
  if (entry.older) {
    entry.older.newer = entry.newer // C. --> E
  }
  entry.newer = undefined // D --x
  entry.older = this.tail // D. --> E
  if (this.tail) {
    this.tail.newer = entry // E. <-- D
  }
  this.tail = entry
  return returnEntry ? entry : entry.value
}

// Export ourselves
if (typeof this === 'object') {
  this.LRUCache = LRUCache
}


index.js
var LRUCache = require('./core').LRUCache

// ----------------------------------------------------------------------------
// Following code is optional and can be removed without breaking the core
// functionality.

/**
 * Check if <key> is in the cache without registering recent use. Feasible if
 * you do not want to chage the state of the cache, but only "peek" at it.
 * Returns the entry associated with <key> if found, or undefined if not found.
 */
LRUCache.prototype.find = function (key) {
  return this._keymap[key]
}

/**
 * Update the value of entry with <key>. Returns the old value, or undefined if
 * entry was not in the cache.
 */
LRUCache.prototype.set = function (key, value) {
  var oldvalue
  var entry = this.get(key, true)
  if (entry) {
    oldvalue = entry.value
    entry.value = value
  } else {
    oldvalue = this.put(key, value)
    if (oldvalue) oldvalue = oldvalue.value
  }
  return oldvalue
}

/**
 * Remove entry <key> from cache and return its value. Returns undefined if not
 * found.
 */
LRUCache.prototype.remove = function (key) {
  var entry = this._keymap[key]
  if (!entry) {
    return
  }
  delete this._keymap[entry.key] // need to do delete unfortunately
  if (entry.newer && entry.older) {
    // relink the older entry with the newer entry
    entry.older.newer = entry.newer
    entry.newer.older = entry.older
  } else if (entry.newer) {
    // remove the link to us
    entry.newer.older = undefined
    // link the newer entry to head
    this.head = entry.newer
  } else if (entry.older) {
    // remove the link to us
    entry.older.newer = undefined
    // link the newer entry to head
    this.tail = entry.older
  } else {// if(entry.older === undefined && entry.newer === undefined) {
    this.head = this.tail = undefined
  }

  this.size--
  return entry.value
}

/** Removes all entries */
LRUCache.prototype.removeAll = function () {
  // This should be safe, as we never expose strong refrences to the outside
  this.head = this.tail = undefined
  this.size = 0
  this._keymap = {}
}

/**
 * Return an array containing all keys of entries stored in the cache object, in
 * arbitrary order.
 */
if (typeof Object.keys === 'function') {
  LRUCache.prototype.keys = function () {
    return Object.keys(this._keymap)
  }
} else {
  LRUCache.prototype.keys = function () {
    var keys = []
    for (var k in this._keymap) keys.push(k)
    return keys
  }
}

/**
 * Call `fun` for each entry. Starting with the newest entry if `desc` is a true
 * value, otherwise starts with the oldest (head) enrty and moves towards the
 * tail.
 *
 * `fun` is called with 3 arguments in the context `context`:
 *   `fun.call(context, Object key, Object value, LRUCache self)`
 */
LRUCache.prototype.forEach = function (fun, context, desc) {
  var entry
  if (context === true) {
    desc = true
    context = undefined
  } else if (typeof context !== 'object') {
    context = this
  }
  if (desc) {
    entry = this.tail
    while (entry) {
      fun.call(context, entry.key, entry.value, this)
      entry = entry.older
    }
  } else {
    entry = this.head
    while (entry) {
      fun.call(context, entry.key, entry.value, this)
      entry = entry.newer
    }
  }
}

/** Returns a JSON (array) representation */
LRUCache.prototype.toJSON = function () {
  var s = []
  var entry = this.head
  while (entry) {
    s.push({key: entry.key.toJSON(), value: entry.value.toJSON()})
    entry = entry.newer
  }
  return s
}

/** Returns a String representation */
LRUCache.prototype.toString = function () {
  var s = ''
  var entry = this.head
  while (entry) {
    s += String(entry.key) + ':' + entry.value
    entry = entry.newer
    if (entry) {
      s += ' < '
    }
  }
  return s
}

// Export ourselves
if (typeof this === 'object') {
  this.LRUCache = LRUCache
}


test-core.js
// Test which will run in nodejs
// $ node test.js
// (Might work with other CommonJS-compatible environments)
'use strict'

var assert = require('assert')
var LRUCache = require('./core').LRUCache
var c = new LRUCache(4)

c.put('adam', 29)
c.put('john', 26)
c.put('angela', 24)
c.put('bob', 48)

assert.equal(c.get('adam'), 29)
assert.equal(c.get('john'), 26)
assert.equal(c.get('angela'), 24)
assert.equal(c.get('bob'), 48)

assert.equal(c.get('angela'), 24)

c.put('ygwie', 81)
assert.equal(c.get('adam'), undefined)

// test remove
var to_remove = new LRUCache(4)
to_remove.put('adam', 29)
to_remove.put('john', 26)
to_remove.shift()
to_remove.shift()
assert.equal(to_remove.size, 0)
assert.equal(to_remove.head, undefined)
assert.equal(to_remove.tail, undefined)

// test shift
c = new LRUCache(4)
assert.equal(c.size, 0)
c.put('a', 1)
c.put('b', 2)
c.put('c', 3)
assert.equal(c.size, 3)
var e = c.shift()
assert.equal(e.key, 'a')
assert.equal(e.value, 1)
e = c.shift()
assert.equal(e.key, 'b')
assert.equal(e.value, 2)
e = c.shift()
assert.equal(e.key, 'c')
assert.equal(e.value, 3)
assert.equal(c.size, 0)

c = new LRUCache(4)

c.put(0, 1)
c.put(0, 2)
c.put(0, 3)
c.put(0, 4)

assert.equal(c.size, 4)
assert.deepEqual(c.shift(), { key: 0, value: 1, newer: undefined, older: undefined })
assert.deepEqual(c.shift(), { key: 0, value: 2, newer: undefined, older: undefined })
assert.deepEqual(c.shift(), { key: 0, value: 3, newer: undefined, older: undefined })
assert.deepEqual(c.shift(), { key: 0, value: 4, newer: undefined, older: undefined })
assert.equal(c.size, 0) // check .size correct
// If we made it down here, all tests passed. Neat.
console.log('All core tests passed!')
